zonemaster/zonemaster-backend
# RPCAPI: fix parameter validation in job_status and user_create with JSON::Validator ⩾ 5.03

## Purpose

This PR fixes a bug where, on systems having JSON::Validator version 5.03 or later installed, calls to some RPCAPI methods like `job_status` or `user_create` failed to have their parameters be validated against their respective schemas. It affects all methods whose schemas are directly constructed by JSON::Validator::Joi and having required parameters, such as `job_status` and `user_create`.

Additionally, we also fix a bug causing a 500 Internal Server Error if we sent the RPCAPI endpoint a string or a number.

The offending code took a `$method_schema`, which could either be a raw Perl hash containing a JSON schema or an opaque object made with JSON::Validator::Joi, and tried to check whether the schema specified any keys as required. It did so by checking whether the `required` key exists in `$method_schema`.

This works fine if `$method_schema` is a Perl hash. But when it’s a JSON::Validator::Joi object, we are touching an implementation detail of that class, which could sooner or later break!

And indeed, it broke: between JSON::Validator version 5.02 and 5.03, an implementation detail was changed, causing issue https://github.com/zonemaster/zonemaster-backend/issues/1105 on systems where JSON::Validator version 5.03 or later is installed. The change seems to have been introduced by commit [788fd42][1] on [JSON::Validator’s Github repo][2].

[1]: https://github.com/jhthorsen/json-validator/commit/788fd4255899339d974d28141a5243f4409067d8
[2]: https://github.com/jhthorsen/json-validator

## Context

Fixes #1105.

## Changes

 * Add unit tests to exercise the `jsonrpc_validate` method in `Zonemaster::Backend::RPCAPI`;
 * Ensure schemas made with JSON::Validator::Joi are compiled before examining them;
 * Add check before attempting to dereference a ref as a hashref.

## How to test this PR

The newly-introduced unit tests should pass.

To test manually:
1. Install JSON::Validator version 5.03 or later on your system;
2. Start Zonemaster::Backend;
3. Perform the following query using curl:
```
curl -s -H "Content-Type: application/json" \
    -d '{ "jsonrpc": "2.0", "method": "job_status", "params": {}, "id": "testing"}' \
     http://localhost:5000/ | jq
```
The output should be equivalent to the following JSON object:
```json
{
  "jsonrpc": "2.0",
  "id": "testing",
  "error": {
    "message": "Invalid method parameter(s).",
    "code": "-32602",
    "data": [
      {
        "path": "/test_id",
        "message": "Missing property"
      }
    ]
  }
}
```
